ts/src/pro/xt.ts

```diff
@@ -0,0 +1,1079 @@
+        //    {
+        //       "topic": "trade",
+        //       "event": "trade@123456",
+        //       "data": {
+        //            "symbol": 'btc_usdt',
+        //            "orderSide": 'SELL',
+        //            "positionSide": 'LONG',
+        //            "orderId": '231485367663419328',
+        //            "price": '27152.7',
+        //            "quantity": '33',
+        //            "marginUnfrozen": '2.85318000',
+        //            "timestamp": 1684892412565
+        //        }
+        //    }
+        //
+        const data = this.safeValue (message, 'data', {});
+        let stored = this.myTrades;
+        if (stored === undefined) {
+            const limit = this.safeInteger (this.options, 'tradesLimit', 1000);
+            stored = new ArrayCacheBySymbolById (limit);
+            this.myTrades = stored;
+        }
+        const parsedTrade = this.parseTrade (data);
+        const market = this.market (parsedTrade['symbol']);
+        stored.append (parsedTrade);
+        const tradeType = market['contract'] ? 'contract' : 'spot';
+        client.resolve (stored, 'trade:' + tradeType);
+    }
+
+    handleMessage (client, message) {
+        const event = this.safeString (message, 'event');
+        if (event === 'pong') {
+            return client.onPong (message);
+        } else if (event !== undefined) {
+            const topic = this.safeString (message, 'topic');
+            const methods = {
+                'kline': this.handleOHLCV,
+                'depth': this.handleOrderBook,
+                'depth_update': this.handleOrderBook,
+                'ticker': this.handleTicker,
+                'agg_ticker': this.handleTicker,
+                'tickers': this.handleTickers,
+                'agg_tickers': this.handleTickers,
+                'balance': this.handleBalance,
+                'order': this.handleOrder,
+            };
+            let method = this.safeValue (methods, topic);
+            if (topic === 'trade') {
+                const data = this.safeValue (message, 'data');
+                if (('oi' in data) || ('orderId' in data)) {
+                    method = this.handleMyTrades;
+                } else {
+                    method = this.handleTrade;
+                }
+            }
+            if (method !== undefined) {
+                return method.call (this, client, message);
+            }
+        }
+    }
+
+    ping (client) {
+        client.lastPong = this.milliseconds ();
+        return 'ping';
+    }
+
+    handleErrorMessage (client: Client, message) {
+        //
+        //    {
+        //        "id": "123",
+        //        "code": 401,
+        //        "msg": "token expire"
+        //    }
+        //
+        const msg = this.safeString (message, 'msg');
+        if ((msg === 'invalid_listen_key') || (msg === 'token expire')) {
+            client.subscriptions['accessToken'] = undefined;
+            this.getListenKey (true);
```

Comment:
why would we get a new key and reject the future immediately after? 🤔

Reply:
What should be at the bottom of the conditional? A `return` statement?

I also don't 100% understand what the purpose of `client.reject` is

Reply:
@samgermain `client.reject` will remove the future from `this.futures` and throw an exception so you should only use it when an error is detected and you want to "abort". In this case, if the process of getting the key was successful yeah I think you can just return

Reply:
> @samgermain `client.reject` will remove the future from `this.futures` and throw an exception so you should only use it when an error is detected and you want to "abort". In this case, if the process of getting the key was successful yeah I think you can just return

What is the future that we're removing? We're rejecting message, which is the object of the error message. I don't really understand what this.futures is. Do we add a future to `this.futures` every time we subscribe to something, or every time we receive a message from the server?